### PR TITLE
fix NPC blocking with broken limbs

### LINKED ISSUE #1790: Stop NPC's from blocking damage using broken limbs
### Is your feature request related to a problem? Please describe.

Problem: Human NPC's block your attacks with their hands, despite them being broken.
Because of that already long fights sometimes can take hours, while both you and the NPC can't do any meaningful damage to each other, despite the log stating that you actually do damage.

### Describe the solution you'd like

Make NPC's (maybe the player too?) unable to block incoming damage using a broken limb. Maybe add some sort of a check that stops the action, if the limb is broken. That's really it.

### Describe alternatives you've considered

Maybe add a feature that makes the limbs get damaged even after they've been broken, straight up destroying it if it receives too much total damage (double the normal amount, maybe?). But, even though this could potentially lead to some interesting ideas and force players to use mutations or bionics to heal lost limbs, and maybe could lead to some interesting enemies that specifically target limbs, it would probably take too much time and effort.

Also, maybe the damage dealt to the broken limb could instead be dealt to the torso or the other limb, that the attack was aimed at, but that would probably create even more bugs.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by leoCottret**
#### Purpose of change
NPCs (and the player) could block with broken limbs if they were wearing an item with the "BLOCK_WHILE_WORN" flag.

#### Describe the solution
Fixes #1790
In the end, this fixes:
- Blocks from BLOCK_WHILE_WORN items still working when all the related limbs are broken (eg both arms for arm guards)
- Blocks not working with MA other than forced unarmed when not using a weapon
- Blocks from forced unarmed still working when both arms are broken

Details:
Some mods can add shields or unusual "BLOCK_WHILE_WORN" worn item, so I tried to take them all into account.

If a worn item covers the arms, it's either an arm guard or shield.
In that case the covered arm must be working (arm guards cover both arms, so one working arm is enough)

Otherwise If a worn item covers the legs, one of them needs to be working

Finally, if an item doesn't cover arms or legs, it skips the health limbs check

#### Testing
Spawn NPC, give him arm guards, hit him, he can block.
Break both his arms, he can't now.
Same for the player.
If an NPC is wielding a weapon with higher blocking, it will still be prioritized for blocking.

### CODE REVIEW

**Comment by chaosvolt** [comment]
Can they also do this bare-handed with enough combat skill, even if lacking armor with `BLOCK_WHILE_WORN`? If so, may be good to tinker with for consistency.

**Comment by leoCottret** [comment]
> Can they also do this bare-handed with enough combat skill, even if lacking armor with `BLOCK_WHILE_WORN`? If so, may be good to tinker with for consistency.

Good call!
The system has a bit of a weird implementation (and some -other- bugs):

If using a MA that is not forced unarmed with blocks, you'll almost never be able to block while unarmed.
To reproduce
- Start a fresh character, set skills to 10, take krav maga. 
- Spawn a zombie, your blocks are working. 
- Then take a weapon and drop it. They don't anymore.
I tried switching styles etc. nothing made the MA blocks work again when unarmed

Then there is the forced unarmed MA, at the opposit, that has working "MA blocks" even when both of your limbs are destroyed.

This is because forced unarmed uses a special boolean.
```cpp
bool force_unarmed = martial_arts_data->is_force_unarmed();
```
And because the check to know if a player unarmed otherwise is
```cpp
bool unarmed = weapon.has_flag( "UNARMED_WEAPON" ); // What about characters without weapons?
```

**TLDR: 
The unarmed variable was always false if the player didn't have a weapon (so the MA blocks didn't work).
The forced unarmed MA blocks were working without weapon because it uses an other special variable**

 But now of course this bug was hiding the other bug:
 Now you can block with broken limbs with any MA that has blocks (yay). So I'll fix that now.

**Comment by leoCottret** [comment]
Ok then the second bug is a bit funny, I thought the MA blocks were already handling broken limbs but I should have take a (much) better look at the code:

```cpp
// the can_leg/arm_block functions check if there's is at least one limb not broken so that means

// if there is at least one leg/arm working, use the left leg or the left arm to block
// note that the left leg/arm can be broken, if the right is working we'll block with the left (what?)
if( martial_arts_data->can_leg_block( *this ) && martial_arts_data->can_arm_block( *this ) ) {
    bp_hit = one_in( 2 ) ? bodypart_id( "leg_l" ) : bodypart_id( "arm_l" );
// otherwise if one of the legs are working, use the left leg (same problem as above??)
} else if( martial_arts_data->can_leg_block( *this ) ) {
    bp_hit = bodypart_id( "leg_l" );
// ok then if all limbs are broken use the left arm (what???)
} else {
    bp_hit = bodypart_id( "arm_l" );
}
// and no there are no extra checks further, this is the part supposed to handle broken limbs
```
So yeah this is why my character now always block with his left arm, with krava maga and 4 broken limbs

File: src/melee.cpp [hunk 1]
```diff
@@ -1645,6 +1659,12 @@ bool Character::block_hit( Creature *source, bodypart_id &bp_hit, damage_instanc
             }
         }
 
+        // if the limb in best condition usable to block is broken, this character can't block anymore
+        if( get_part_hp_cur( bp_hit ) <= 0 ) {
+            blocks_left = 0;
```
Comment (Coolthulhu) [inline src/melee.cpp:1664]: Why set `blocks_left` to 0?

Comment (leoCottret) [reply src/melee.cpp:1664]: In this case that means our character has either his 2 arms broken with a MA that allows arms block.
Or his 4 limbs broken with a MA that allows legs blocks.
In either case he can't hold a weapon anymore.
If he had a magic item that could still block, has_shield would be true and this part would be used instead
`if( !force_unarmed && has_shield ) {`

So it's a slight optimisation to avoid entering in Character::block_hit again
Cf here:
```cpp
if( blocks_left < 1 ) {
      return false;
  }

blocks_left--;
```
You want me to remove it noneless?

Comment (Coolthulhu) [reply src/melee.cpp:1664]: Would be good to explain the reasoning in the comment, in case someone else looks at it and decides to rewrite it.
For example, summarize assumptions.
> At this point, we know we won't try blocking with items, only with limbs. But there are no limbs left, so we can disable further attempts at blocking.

Or otherwise explain that the next blocking check would necessarily have to end up here, so disabling blocks only skips the "loop".

Comment (leoCottret) [reply src/melee.cpp:1665]: Yes I think your sentence summarize well what I said above, I added it as comment.
